spec/solr_indexer_spec.rb

```diff
@@ -1,74 +1,90 @@
 # frozen_string_literal: true
 require 'spec_helper'
+require 'find'
+require 'gingr/solr_indexer'
 
 RSpec.describe Gingr::SolrIndexer do
-  let(:url) { 'http://solr:8983/solr/geodata-test' }
-  let(:solr) { instance_double(RSolr::Client) }
+  around(:each) do |test|
+    original_solr_url = ENV['SOLR_URL']
+    test.run
+  ensure
+    ENV['SOLR_URL'] = original_solr_url
+  end
 
   describe '#initialize' do
-    before do
-      allow(RSolr).to receive(:connect).and_return(solr)
+    it 'initializes a solr client with the given url' do
+      indexer = Gingr::SolrIndexer.new('http://solr-from-init/')
+      expect(indexer.solr.uri.to_s).to eq 'http://solr-from-init/'
+    end
+
+    it 'falls back to ENV["SOLR_URL"] if it is set' do
+      ENV['SOLR_URL'] = 'http://solr-from-env/'
+      indexer = Gingr::SolrIndexer.new
+      expect(indexer.solr.uri.to_s).to eq 'http://solr-from-env/'
     end
 
-    it 'should initialize' do
-      described_class.new(url)
-      expect(RSolr).to have_received(:connect).with(
-        url:,
-        adapter: :net_http_persistent
-      )
+    it 'falls back to the config' do
+      ENV.delete 'SOLR_URL'
+      indexer = Gingr::SolrIndexer.new
+      expect(indexer.solr.uri.to_s).to eq 'http://solr:8983/solr/geodata-test/'
     end
   end
 
-  describe '#update' do
-    let(:file_path) { 'spec/fixture/jsonfile/berkeley_public_pdf.json' }
-    let(:doc) { JSON.parse(File.read(file_path)) }
+  describe '#update_reference_urls!' do
+    let(:document) { JSON.load_file('spec/fixture/jsonfile/berkeley_public_pdf.json') }
 
-    before do
-      allow(solr).to receive(:update)
-      allow(RSolr).to receive(:connect).and_return(solr)
-      solr_indexer.update(file_path)
+    it 'does nothing if reference_urls are nil' do
+      indexer = Gingr::SolrIndexer.new
+      expect { indexer.update_reference_urls! document }.not_to change { document }
```

Comment:
The [change matcher](https://rubydoc.info/gems/rspec-expectations/RSpec%2FMatchers:change) is pretty nifty! It asserts that the value of the block doesn't change before and after the expected block is run.

Reply:
This matcher is nity.